refactor(NfHandler) replace TowerAnalysisAPI and NextflowAnalysisAPI

fevac opened the pull request:
## Description

Part  of https://github.com/Clinical-Genomics/cg/pull/2156

Refactor of TowerAnalysisAPI and NextflowAnalysisAPI. Moved methods to keep class scope.
Another PR will follow up and further modify these classes to clean up common NF code.


### Changed

- Removed TowerAnalysisAPI and NextflowAnalysisAPI and replaced by NfHandler NfTowerHandler and NextflowHandler



### How to prepare for test

- [x] Ssh to relevant server (depending on type of change)
- [x] Use stage: `us`
- [x] Paxa the environment: `paxa`
- [x] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_main -t cg -b refactor_nf_handler -a
    ```

### How to test

- [x] `cg workflow rnafusion start <case_id>` : case should start in tower
- [x] `cg workflow rnafusion store <case_id>`: should be succesful


## Review

- [x] Tests executed by EFC
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

fevac commented:
@ivadym it would be great if you could review this and keep focus on the new class names and if their methods are well placed. Thanks! ❤️

ChrOertlin commented on cg/meta/workflow/nf_analysis.py:
```diff
@@ -63,43 +66,103 @@ def get_workflow_manager(self) -> str:
-        """Return a path where the <case>_metrics_deliverables.yaml file should be located."""
-        return Path(self.root_dir, case_id, f"{case_id}_metrics_deliverables").with_suffix(
-            FileExtensions.YAML
+    def get_replace_map(self, case_id: str) -> dict:
+        """Get a mapping to replace constants from template to create case deliverables."""
+        return {
+            "PATHTOCASE": str(self.get_case_path(case_id)),
+            "CASEID": case_id,
+        }
```
Is it possible to use a data model here instead of passing a dictionary?
```suggestion
    def get_replace_map(self, case_id: str) -> dict:
        """Get a mapping to replace constants from template to create case deliverables."""
        return {
            "PATHTOCASE": str(self.get_case_path(case_id)),
            "CASEID": case_id,
        }
```

ChrOertlin commented on cg/meta/workflow/rnafusion.py:
```diff
@@ -93,6 +89,19 @@ def build_samplesheet_content(
+        headers: List[str],
+        config_path: Path,
+    ) -> None:
+        """Write sample sheet csv file."""
+        with open(config_path, "w") as outfile:
+            outfile.write(",".join(headers))
+            for i in range(len(samplesheet_content[NFX_SAMPLE_HEADER])):
+                outfile.write("\n")
+                outfile.write(",".join([samplesheet_content[k][i] for k in headers]))
```
We have a csv write controller in io that we should use here

fevac replied:
All the sample sheets methods need to be cleaned up. They are quite messy. I was hoping to get away with them on this PR. But FINE I can do it here 😬

fevac replied:
oh I hadn't read properly 😶‍🌫️  leaving this for a following PR

ChrOertlin commented on cg/meta/workflow/rnafusion.py:
```diff
@@ -93,6 +89,19 @@ def build_samplesheet_content(
         }
         return samplesheet_content
 
+    def create_samplesheet_csv(
```
```suggestion
    def write_samplesheet_csv(
```
This function seems to not create but write the sample sheet

ivadym replied:
Static method

ChrOertlin commented on cg/meta/workflow/rnafusion.py:
```diff
@@ -158,12 +159,8 @@ def get_references_path(self, genomes_base: Optional[Path] = None) -> Path:
-            "outdir": NextflowAnalysisAPI.get_outdir_path(
-                case_id=case_id, root_dir=self.root_dir
-            ).as_posix(),
+            "input": self.get_case_path(case_id=case_id).as_posix(),
+            "outdir": self.get_case_path(case_id=case_id).as_posix(),
             "genomes_base": self.get_references_path().as_posix(),
             "trim": RnafusionDefaults.TRIM,
             "fastp_trim": RnafusionDefaults.FASTP_TRIM,
```
same as above, but arguable too much to change when if moving away from dict to models, maybe something for the future

fevac replied:
in a following PR

ivadym commented on cg/utils/nf_handlers.py:
```diff
@@ -0,0 +1,217 @@
+from cg.utils.utils import build_command_from_dict
+
+
+class NfHandler:
```
I agree. And I have the same feeling about `NfAnalysisAPI`. To me at least, `NfCoreAnalysisAPI` appears to be more comprehensible.

fevac replied:
I like the `NfBaseHandler`, however I'm hesitant about `NfCoreAnalysisAPI` as those pipelines don't need to be NF-Core pipelines, just written in nextflow (independently of they way they are run).

fevac replied:
updated to NfBaseHandler

ivadym commented on cg/cli/workflow/rnafusion/base.py:
```diff
@@ -114,22 +114,22 @@ def run(
 
     command_args: CommandArgs = CommandArgs(
```
After this comment and going through the PR, my logic is telling me than all these handler methods belong to the `NfAnalysisAPI`, as they are directly related to the analysis. Additionally, there are certain methods, such as `get_case_path`, that appear to share functionalities.

fevac replied:
moved

ivadym commented on cg/meta/workflow/rnafusion.py:
```diff
@@ -215,26 +210,26 @@ def run_analysis(
                 environment=self.conda_env,
                 conda_binary=self.conda_binary,
-                launch_directory=NextflowAnalysisAPI.get_case_path(
+                launch_directory=NextflowHandler.get_case_path(
```
```suggestion
                launch_directory=self.get_case_path(
```

ivadym reviewed (commented):
Great work! It's getting there 🚀 🌕 👩‍🚀

henrikstranneheim commented on cg/utils/nf_handlers.py:
```diff
@@ -0,0 +1,217 @@
+        return nextflow_options + ["run", pipeline_path] + run_options
+
+    @classmethod
+    def get_log_path(cls, case_id: str, pipeline: str, root_dir: str, log: str = None) -> Path:
```
Add doc str

henrikstranneheim commented on cg/utils/nf_handlers.py:
```diff
@@ -0,0 +1,217 @@
+
+    @classmethod
+    def get_sbatch_path(cls, case_id: str, root_dir: str) -> Path:
+        """Returns a path where the nextflow sbatch for the head job should be located."""
```
```suggestion
        """Returns a path where the Nextflow sbatch for the head job should be located."""
```

henrikstranneheim commented on cg/utils/nf_handlers.py:
```diff
@@ -0,0 +1,217 @@
+    def get_sbatch_path(cls, case_id: str, root_dir: str) -> Path:
+        """Returns a path where the nextflow sbatch for the head job should be located."""
+        return Path(
+            cls.get_case_path(case_id=case_id, root_dir=root_dir), "nextflow_head_job.sbatch"
```
Use FileExtension constant

henrikstranneheim commented on cg/utils/nf_handlers.py:
```diff
@@ -0,0 +1,217 @@
+        number_tasks: int = SlurmHeadJobDefaults.NUMBER_TASKS,
+        dry_run: bool = False,
+    ) -> int:
+        """Executes nextflow head job command."""
```
```suggestion
        """Executes Nextflow head job command."""
```

fevac commented:
I have renamed `NfHandler` to `NfBaseHandler` and moved away some methods that didn't belong. NfBaseHandler should have methods and attributes that are common to both Nextflow and Tower. These handlers should deal with setting parameters and setting the execution (which is different when running directly with nextflow  or through tower). These handlers and their methods will be further refactored in a coming PR. The idea is that NF pipelines (rnafusion, taxprofiler) can be easily run with either Nextflow or Tower by setting a property in their respective classes.

I've cleaned up the code base on your review but I don't want to make this PR massive. So I'm leaving out, for coming PRs:
- Using data model for `get_replace_map` https://github.com/Clinical-Genomics/cg/pull/2309#discussion_r1288072632 and https://github.com/Clinical-Genomics/cg/pull/2309#discussion_r1288074221
- use csv write controller and refactor sample sheets methods: https://github.com/Clinical-Genomics/cg/pull/2309#discussion_r1288075216 and https://github.com/Clinical-Genomics/cg/pull/2309#discussion_r1288247756
- datamodel for rnafusion get_default_parameters https://github.com/Clinical-Genomics/cg/pull/2309#discussion_r1288079329
- data model for metadata https://github.com/Clinical-Genomics/cg/pull/2309#discussion_r1288234759

ivadym commented on cg/utils/nf_handlers.py:
These handlers, although they perform utility functions, appear to me to be more tailored to the Analysis API rather than to the general-purpose utilities. I would suggest moving this file to the same level as the `FastqHandler`, within the `meta/workflow` directory.

fevac replied:
good point! I'll move them

fevac replied:
well... thinking about it, should it really be within workflow? is not really a workflow... Not sure if I like it

ivadym replied:
I think it's part of the workflow domain, similar to `FastqHandler`... you could also create a `workflow_handlers` folder in `meta`. But `utils` are used for more general-purpose utility functions, and these classes might not fully align with that

ivadym replied:
if you want to think about it and addressed it a future PR it's also fine 😄

fevac replied:
No... I'd like to be done with it. I'll move it as suggested to meta/workflow

ivadym reviewed (approved):
Nice refactoring ⭐  Last comment from my side 😋

fevac commented on cg/meta/workflow/nf_analysis.py:
```diff
@@ -63,43 +66,103 @@ def get_workflow_manager(self) -> str:
+        else:
+            raise ValueError("Either forward or reverse needs to be specified")
+        sorted_metadata: list = sorted(metadata, key=operator.itemgetter("path"))
+        return [d["path"] for d in sorted_metadata if d["read"] == read_number]
```
yes! in a following PR!

fevac commented:
tests passing

fevac commented:
- Deployed in stage:
<img width="543" alt="image" src="https://github.com/Clinical-Genomics/cg/assets/29628428/214d49b2-59ee-4b41-92af-72970e72380e">


- Deployed in prod:
<img width="497" alt="image" src="https://github.com/Clinical-Genomics/cg/assets/29628428/21197303-dc25-4da2-9767-7ac19bb9f2c8">
